pkg/rabbit/connections_handler.go

```diff
@@ -0,0 +1,247 @@
+	CloseRabbitMQConnections()
+}
+
+type RabbitMQConnectionInterface interface {
+	NotifyClose(chan *amqp.Error) chan *amqp.Error
+	Close() error
+	IsClosed() bool
+}
+
+type RabbitMQConnectionWrapperInterface interface {
+	RabbitMQConnectionInterface
+	ChannelWrapper() (RabbitMQChannelInterface, error)
+}
+
+type RabbitMQChannelInterface interface {
+	IsClosed() bool
+	NotifyClose(chan *amqp.Error) chan *amqp.Error
+	Qos(int, int, bool) error
+	Confirm(bool) error
+	Consume(string, string, bool, bool, bool, bool, amqp.Table) (<-chan amqp.Delivery, error)
+	PublishWithDeferredConfirm(string, string, bool, bool, amqp.Publishing) (*amqp.DeferredConfirmation, error)
+	QueueInspect(string) (amqp.Queue, error)
+}
+
+type RabbitMQHelper struct {
+	firstSetup    bool
+	cycleDuration time.Duration
+	Connection    RabbitMQConnectionWrapperInterface
+	Channel       RabbitMQChannelInterface
+
+	logger *zap.SugaredLogger
+}
+
+type RabbitMQConnection struct {
+	connection interface{}
+}
+
+func NewConnection(conn interface{}) *RabbitMQConnection {
+	return &RabbitMQConnection{connection: conn}
+}
+
+func (r *RabbitMQConnection) ChannelWrapper() (RabbitMQChannelInterface, error) {
+	if c, ok := r.connection.(*amqp.Connection); ok {
+		return c.Channel()
+	} else if ci, ok := r.connection.(RabbitMQConnectionWrapperInterface); ok {
+		return ci.ChannelWrapper()
+	}
+	return nil, errors.New("wrong typed connection arg")
+}
+
+func (r *RabbitMQConnection) NotifyClose(c chan *amqp.Error) chan *amqp.Error {
+	if ci, ok := r.connection.(RabbitMQConnectionInterface); ok {
+		return ci.NotifyClose(c)
+	}
+	close(c)
+	return nil
+}
+
+func (r *RabbitMQConnection) Close() error {
+	if ci, ok := r.connection.(RabbitMQConnectionInterface); ok {
+		return ci.Close()
+	}
+	return errors.New("wrong typed connection arg")
+}
+
+func (r *RabbitMQConnection) IsClosed() bool {
+	if ci, ok := r.connection.(RabbitMQConnectionInterface); ok {
+		return ci.IsClosed()
+	}
+	return true
+}
+
+func NewRabbitMQHelper(cycleDuration time.Duration, logger *zap.SugaredLogger) RabbitMQConnectionsHandlerInterface {
+	return &RabbitMQHelper{
+		firstSetup:    true,
+		cycleDuration: cycleDuration,
+		logger:        logger,
+	}
+}
+
+func (r *RabbitMQHelper) SetupRabbitMQConnectionAndChannel(
+	ctx context.Context,
+	rabbitMQURL string,
+	configFunction func(RabbitMQConnectionInterface, RabbitMQChannelInterface) error,
+	dialFunc func(string) (RabbitMQConnectionWrapperInterface, error)) {
+	var err error
+	retryConnection, retryChannel := true, true
+	for retryConnection || retryChannel {
+		// Wait one cycle duration always except the first time
+		if !r.firstSetup {
+			time.Sleep(time.Millisecond * r.cycleDuration)
+		} else {
+			r.logger.Info("Creating and Configuring RabbitMQ Connection and Channel")
+			r.firstSetup = false
+		}
+		r.Connection, r.Channel, err = r.CreateAndConfigConnectionsAndChannel(&retryConnection, &retryChannel, rabbitMQURL, configFunction, dialFunc)
```

Comment:
The way I understand it, the retry booleans save us from recreating a connection when a connection was successful but the channel creation failed. 

It feels overly complex just to save from that one case. It might make more sense to just rely on the `err`. If err happened, call `CloseRabbitMQConnections` just for clean up and then retry after the 1 second wait.